### ISSUE TITLE
Conditional in Database\Transaction:shouldBeginTransaction() my have an extra clause

### ISSUE THREAD

**Issue by kdebisschop** [wizaplace/php-etl#76]
After refactoring for PHP 7.4, the current function shouldBeginTransaction() in \Wizaplace\Etl\DatabaseTransaction is:

`return !$this->open && (0 === $this->size || 1 === $this->count);`

It's not clear to me that `0 === $this->size` should be part of that condition and in fact

`return !$this->open && 1 === $this->count;`

also passes all our current tests. Either the tests are not diagnostic or the code is inaccurate, We should determine which and resolve the inconsistency.

**Comment by kdebisschop** [comment]
Sure enough, I ran infection against the code and if confirms what I saw from my manual test - although the line of code are covered in the sense they are executed, the test is not diagnostic of any possible problems with the role of $size in thta conditional. From the infection log:

> --- Original
> +++ New
> @@ @@
>       */
>      protected function shouldBeginTransaction() : bool
>      {
> -        return !$this->open && (0 === $this->size || 1 === $this->count);
> +        return !$this->open && (-1 === $this->size || 1 === $this->count);
>      }
>      /**
>       * Check if it should commit a transaction.
> 
> 
> 15) php-etl/src/Database/Transaction.php:85    [M] OneZeroInteger
> 
> --- Original
> +++ New
> @@ @@
>       */
>      protected function shouldBeginTransaction() : bool
>      {
> -        return !$this->open && (0 === $this->size || 1 === $this->count);
> +        return !$this->open && (1 === $this->size || 1 === $this->count);
>      }
>      /**
>       * Check if it should commit a transaction.
> 
> 
> 16) php-etl/src/Database/Transaction.php:85    [M] Identical
> 
> --- Original
> +++ New
> @@ @@
>       */
>      protected function shouldBeginTransaction() : bool
>      {
> -        return !$this->open && (0 === $this->size || 1 === $this->count);
> +        return !$this->open && (0 !== $this->size || 1 === $this->count);
>      }
>      /**
>       * Check if it should commit a transaction.
>
>
> 17) php-etl/src/Etl.php:105    [M] FalseValue


BTW - what infection does for us is really cool - I had not been aware of the tool until contributing to this project. I'm not sure there is a way to integrate it into CI because it can generate false positives. But we probably should be using it more, and/or making our use of it explicit.

**Comment by Nicolas-Masson-Wizaplace** [comment]
> BTW - what infection does for us is really cool - I had not been aware of the tool until contributing to this project. I'm not sure there is a way to integrate it into CI because it can generate false positives. But we probably should be using it more, and/or making our use of it explicit.

Sure, at least, a little mutation badge would be really cool.
https://infection.github.io/guide/mutation-badge.htm

**Comment by Nicolas-Masson-Wizaplace** [comment]
@ecourtial 
What do you think about add infection to the CI with a reasonable min-score ?
We have pretty high score so far

```
Metrics:
         Mutation Score Indicator (MSI): 77%
         Mutation Code Coverage: 90%
         Covered Code MSI: 85%
```

**Comment by ecourtial** [comment]
As a huge fan off infection, I would say yes :100: !  We should add the codecov and infection cov badges. I did it for one of my personal project ([badge](https://github.com/ecourtial/Fighting-Steel-Dynamic-Campaign-Manager) and CI [job](https://github.com/ecourtial/Fighting-Steel-Dynamic-Campaign-Manager/blob/05b5382f395357424d7de2c06d7a1a3ebedc226b/.circleci/config.yml#L103)).

At the beginning, we did not paid to much attention to infection score in this project because it was very low and would have needed a lot of refacto. Could be interesting to see what we got after we merge #71 .
